### ISSUE TITLE
recall and prediction different in model.evaluate than manual calculation

### ISSUE THREAD

**Issue by tuxmania87** [keras-team/keras#14060]
i have a Keras Sequential Network and use the validation_data attribute to get an idea of the validation straight away. 

However when the model is trained and I predict on the validation set I dont get as high scores as keras is telling me through training.

This is the model and training:

```
def getKerasModel(ndim):

    model = Sequential()
    model.add(Dense(100, activation='relu', input_shape=(ndim,)))
    model.add(Dense(40, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(optimizer="adam", loss="binary_crossentropy", metrics=[keras.metrics.Precision(), keras.metrics.Recall()])
    return model
```
Then I train it:
```
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import to_categorical

# pipeline is doing scales and one hot encoding
X_train2 = full_pipeline.fit_transform(X_train)
X_val2 = full_pipeline.transform(X_val)

model = getKerasModel(ndim=X_train2.shape[1])
model.fit(X_train2, y_train,epochs=5, batch_size=32, verbose=True, validation_data = (X_val2, y_val))
```
Now this is the result:
```
Train on 265815 samples, validate on 38247 samples
Epoch 1/5
265815/265815 [==============================] - 12s 45us/step - loss: 0.2715 - precision: 0.6690 - recall: 0.4065 - val_loss: 0.2965 - val_precision: 0.6875 - val_recall: 0.4654
Epoch 2/5
265815/265815 [==============================] - 13s 50us/step - loss: 0.2300 - precision: 0.6956 - recall: 0.4959 - val_loss: 0.3735 - val_precision: 0.7000 - val_recall: 0.5192
Epoch 3/5
265815/265815 [==============================] - 13s 47us/step - loss: 0.2111 - precision: 0.7020 - recall: 0.5376 - val_loss: 0.4003 - val_precision: 0.7071 - val_recall: 0.5509
Epoch 4/5
265815/265815 [==============================] - 17s 63us/step - loss: 0.1988 - precision: 0.7113 - recall: 0.5627 - val_loss: 0.4695 - val_precision: 0.7113 - val_recall: 0.5735
Epoch 5/5
265815/265815 [==============================] - 16s 61us/step - loss: 0.1900 - precision: 0.7120 - recall: 0.5831 - val_loss: 0.4893 - val_precision: 0.7134 - val_recall: 0.5908
```
So it definetly says after the last epoch :
```
val_precision: 0.7134 - val_recall: 0.5908
```
When I now use the model, do prediction and evaluate precision and recall seperatly it is much lower:
```
pr = model.predict(X_val2)

binary_result = [1 if i[0] > 0.5 else 0 for i in pr]
print(precision_score(y_val,binary_result))
print(recall_score(y_val,binary_result))

>> 0.23
>> 0.38
```
I don't get why keras is doing much better in its evaluation

**Comment by U0M0Z** [comment]
I have the same issue; a model where Keras evaluate score is at 90% and the overall recall_score from scikit_learn at 70%. 

The difference is really not clear to me

**Comment by rohanshad** [comment]
Same issue here but with keras.metrics.auc, I believe the underlying issue is the same as it affects precision & recall too. The auc using `model.evaluate()` is consistently 10-30% higher than that produced by scikit_learn or the pROC package in R using the `model.predict()` probabilities on the exact same dataset with shuffling disabled.

**Comment by edmondja** [comment]
I have the same issue : 
`validX = [df2[states].loc[indices['valid']]*0, df2[actions].loc[indices['valid']]*0]
print(validX)
target = df2[rewards].loc[indices['valid']].mean(axis=1)*100*0
print('target')
print(target)
print('prediction')
print(critic.predict(validX)[:, 0])
print('manual mse')
print(np.mean((critic.predict(validX)[:, 0] - target.values)**2))
print('evaluation')
critic.evaluate(validX, target.values, verbose=1)
[      state_0  state_1  state_2  ...  state_21  state_22  state_23
9912     -0.0      0.0     -0.0  ...       0.0       0.0       0.0
9913      0.0     -0.0      0.0  ...       0.0       0.0       0.0
9918      0.0     -0.0      0.0  ...       0.0       0.0       0.0

[3 rows x 24 columns],       action_0  action_1  action_2  action_3
9912       0.0       0.0      -0.0       0.0
9913      -0.0       0.0       0.0      -0.0
9918      -0.0       0.0       0.0      -0.0]
target
9912   -0.0
9913   -0.0
9918   -0.0
dtype: float64
prediction
[-4.979239 -4.979239 -4.979239]
manual mse
24.792820889352015
evaluation
44/44 [==============================] - 0s 632us/step - loss: 5.0198
Out[82]: 5.0198211669921875

validX = [df2[states].loc[indices['valid']]*0+1, df2[actions].loc[indices['valid']]*0+1]
print(validX)
target = df2[rewards].loc[indices['valid']].mean(axis=1)*100*0
print('target')
print(target)
print('prediction')
print(critic.predict(validX)[:, 0])
print('manual mse')
print(np.mean((critic.predict(validX)[:, 0] - target.values)**2))
print('evaluation')
critic.evaluate(validX, target.values, verbose=1)
[      state_0  state_1  state_2  ...  state_21  state_22  state_23
9912      1.0      1.0      1.0  ...       1.0       1.0       1.0
9913      1.0      1.0      1.0  ...       1.0       1.0       1.0
9918      1.0      1.0      1.0  ...       1.0       1.0       1.0

[3 rows x 24 columns],       action_0  action_1  action_2  action_3
9912       1.0       1.0       1.0       1.0
9913       1.0       1.0       1.0       1.0
9918       1.0       1.0       1.0       1.0]
target
9912   -0.0
9913   -0.0
9918   -0.0
dtype: float64
prediction
[-3.1966157 -3.1966157 -3.196616 ]
manual mse
10.218352415703842
evaluation
44/44 [==============================] - 0s 659us/step - loss: 5.0198
Out[83]: 5.0198211669921875`

No matter what is my data, evaluate always returns the same mse.
The problem appears too with theano backend. 
Its kind of shocking, it makes me wonder if we can even trust keras in properly backpropagating.

**Comment by rohanshad** [comment]
For my problem I realized it was to do with the way we were assigning labels. For a binary classification problem 1-hot categorical vectors + softmax output led to the strange effects we were seeing. We converted this to a 1-D float label, and used a sigmoid output and everything works as expected now. 

I'm not sure if this helps your case, but I'd check the labels + output layer.

**Comment by edmondja** [comment]
Keras 2.2.5 with Tensorflow 1.15.4 works fine for me.

**Comment by sushreebarsa** [comment]
@tuxmania87 Moving this issue to closed status as there has been no recent activity.In case you still face the error please create a new issue,we will get you the right help.Thanks!
